main/lsp/requests/document_symbol.cc

```diff
@@ -12,174 +13,172 @@ using namespace std;
+                this->scope->children.emplace_back(std::make_shared<Node>(this->scope, child, childLoc));
+                return;
+            }
+        }
     }
 
-    for (auto mem : sym.asClassOrModuleRef().data(gs)->membersStableOrderSlow(gs)) {
-        if (mem.first == core::Names::attached() || mem.first == core::Names::singleton()) {
-            continue;
+public:
+    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &expr) {
+        auto &klass = ast::cast_tree_nonnull<ast::ClassDef>(expr);
+        if (klass.symbol == core::Symbols::root()) {
+            return;
         }
-        if (!absl::c_any_of(mem.second.locs(gs), [&filter](const auto &loc) { return loc.file() == filter; })) {
-            if (!forced.contains(mem.second)) {
+
+        auto next = std::make_shared<Node>(this->scope, klass.symbol, ctx.locAt(klass.loc));
+        next->declLoc = ctx.locAt(klass.declLoc);
+        this->scope->children.emplace_back(next);
+        this->scope = std::move(next);
+
+        for (auto member : klass.symbol.data(ctx)->members()) {
+            if (member.second.isClassOrModule() || member.second.isMethod()) {
                 continue;
             }
+
+            this->addChild(ctx, member.second);
         }
-        auto inner = symbolRef2DocumentSymbol(gs, mem.second, filter, defMapping, forced);
-        if (inner) {
-            out.push_back(move(inner));
-        }
-    }
-}
 
-struct RangeInfo {
-    unique_ptr<Range> range;
-    unique_ptr<Range> selectionRange;
-};
+        auto singleton = klass.symbol.data(ctx)->lookupSingletonClass(ctx);
+        if (singleton.exists()) {
+            for (auto member : singleton.data(ctx)->members()) {
+                if (member.second.isClassOrModule() || member.second.isMethod()) {
+                    continue;
+                }
 
-std::optional<RangeInfo> rangesForSymbol(const core::GlobalState &gs, core::SymbolRef symRef,
-                                         const UnorderedMap<core::SymbolRef, ASTSymbolInfo> &defMapping,
-                                         const UnorderedMap<core::SymbolRef, core::SymbolRef> &forced) {
-    RangeInfo info;
-    auto it = defMapping.find(symRef);
-    if (it != defMapping.end()) {
-        info.range = Range::fromLoc(gs, it->second.loc);
-        info.selectionRange = Range::fromLoc(gs, it->second.declLoc);
-    } else {
-        auto it = forced.find(symRef);
-        if (it != forced.end()) {
-            return rangesForSymbol(gs, it->second, defMapping, forced);
+                this->addChild(ctx, member.second);
+            }
         }
+    }
 
-        auto loc = symRef.loc(gs);
-        if (!loc.file().exists()) {
-            return nullopt;
+    void postTransformClassDef(core::Context ctx, ast::ExpressionPtr &expr) {
+        auto &klass = ast::cast_tree_nonnull<ast::ClassDef>(expr);
+        if (klass.symbol == core::Symbols::root()) {
```

Comment:
Why is `root` special here?

Reply:
(It's probably fine to be special, I'm just curious to hear context.)

Reply:
We eagerly add it when constructing the outliner, and special casing it here avoids accidentally popping that entry and setting `this->scope` to `nullptr`.